vault-core/src/main/java/org/apache/jackrabbit/vault/fs/io/DocViewFormat.java

```diff
@@ -0,0 +1,215 @@
+            return false;
+        }
+    }
+
+    /**
+     * Checks if given files are in the format returned by {@link DocViewFormat#getXmlOutputFormat()} by traversing the directory tree
+     * below the given file. Only files that match at least one of the given filenamePatterns are checked.
+     *
+     * @param file             the given directory or file
+     * @param filenamePatterns a list of {@link Pattern}s applied again the file's names while traversing.
+     * @return as list of relative paths of those files which are not formatted correctly according to {@link DocViewFormat#checkFormat(File)}
+     * @throws IOException in case there is an exception during traversal
+     */
+    public static List<String> checkFormat(File file, List<Pattern> filenamePatterns) throws IOException {
+        final List<String> malformedFiles = new LinkedList<>();
+        Files.walkFileTree(file.toPath(), new AbstractFormattingVisitor(filenamePatterns) {
+            @Override protected void process(File file) {
+                if (!checkFormat(file)) {
+                    malformedFiles.add(file.getPath());
+                }
+            }
+        });
+
+        return malformedFiles;
+    }
+
+    /**
+     * Formats a given file using the {@link OutputFormat} returned by {@link DocViewFormat#getXmlOutputFormat()}. The file is replaced on
+     * disk but only if wasn't already formatted correctly.
+     *
+     * @param file
+     * @throws IOException
+     */
+    public static void format(File file) throws IOException {
+        CRC32 originalCrc32 = new CRC32();
+        CRC32 formattedCrc32 = new CRC32();
+        byte[] formatted = format(file, originalCrc32, formattedCrc32);
+
+        if (originalCrc32.getValue() != formattedCrc32.getValue()) {
+            try (OutputStream out = new BufferedOutputStream(new FileOutputStream(file))) {
+                IOUtils.copy(new ByteArrayInputStream(formatted), out);
+            }
+        }
+    }
+
+    /**
+     * Formats given files using the {@link OutputFormat} returned by {@link DocViewFormat#getXmlOutputFormat()} by traversing the directory
+     * tree given as file. Only those files will be formatted, that have a filename matching at least one of the given filenamePatterns.
+     *
+     * @param file
+     * @param filenamePatterns
+     * @throws IOException in case there is an exception during traversal or formatting. That means formatting will fail on the first error that appeared
+     */
+    public static void format(File file, List<Pattern> filenamePatterns) throws IOException {
+        Files.walkFileTree(file.toPath(), new AbstractFormattingVisitor(filenamePatterns) {
+            @Override protected void process(File file) throws IOException {
+                format(file);
+            }
+        });
+    }
+
+    private static synchronized byte[] format(File file, Checksum original, Checksum formatted) throws IOException {
```

Comment:
why synchronized ?

Reply:
Mainly because of the buffer reused for formatted files. I made the methods non-static which makes sense for concurrency but also due to the fact that OutputFormat is mutable and though should not be hold statically.